### PR TITLE
#2932 Create new recipe from existing extension in page editor

### LINKED ISSUE #2932: Create a new recipe in page editor from existing personal extension
Related to https://github.com/pixiebrix/pixiebrix-extension/issues/2557

Designs are "idea #2" here: https://balsamiq.cloud/s7ffzg5/pu105es/r344E

- Add action button to the sidebar header that appears when selecting a personal extension
- On click, show a new-recipe-configuration modal
- On modal "save", create the recipe and attach the extension

### PR DESCRIPTION
**Pull request by BLoe**
Closes #2932

### CODE REVIEW

**Comment by BLoe** [comment]
Commits specific to this PR/branch start with [2f6e0ad](https://github.com/pixiebrix/pixiebrix-extension/pull/3170/commits/2f6e0ad863c3da35cfdd9236c84d49cad4ac41d9), so it's just the 4 commits there at the end.

**Comment by BALEHOK** [comment]
> Commits specific to this PR/branch start with [2f6e0ad](https://github.com/pixiebrix/pixiebrix-extension/pull/3170/commits/2f6e0ad863c3da35cfdd9236c84d49cad4ac41d9), so it's just the 4 commits there at the end. This is the last one that will be like this, promise.

You could take those 4 commits and rebase onto main.

**Comment by BLoe** [comment]
>rebase onto main

This doesn't quite work because the previous commits don't exist in main, they were squashed into one merge commit when  each previous PR was merged. So if I rebase it usually tries to go way back and rebase all the commits and I get tons of conflicts I have to resolve anyway. In the past sometimes I'll make a new branch and cherry-pick over the commits, but I already pushed this branch. I can't do that when I want to put up the draft ahead of time. I'm going to rebase onto `main` for the next PR before I push the branch.

File: src/pageEditor/panes/save/saveHelpers.ts [hunk 1]
```diff
@@ -256,40 +269,50 @@ export function replaceRecipeExtension(
...
 }: RecipeParts): UnsavedRecipeDefinition {
-  const cleanRecipeExtensions = installedExtensions.filter(
-    (extension) =>
-      extension._recipe?.id === sourceRecipe.metadata.id &&
-      !dirtyRecipeElements.some((element) => element.uuid === extension.id)
-  );
-
-  return produce(sourceRecipe, (draft) => {
+  // If there's no source recipe, then we're creating a new one, so instead we start
+  // with a blank slate that will be filled in
```
Comment (BALEHOK) [inline src/pageEditor/panes/save/saveHelpers.ts:299]: "a blank slate" - typo?

File: src/pageEditor/sidebar/AddToRecipeModal.tsx [hunk 2]
```diff
@@ -26,36 +26,51 @@ import {
...
   const hideModal = () => {
     dispatch(editorActions.hideAddToRecipeModal());
   };
 
-  function onConfirmAddToRecipe(
-    recipeMetadata: RecipeMetadata,
-    keepLocalCopy: boolean
-  ) {
+  function onConfirmAddToRecipe(recipeId: RegistryId, keepLocalCopy: boolean) {
+    if (recipeId === "@new") {
```
Comment (BALEHOK) [inline src/pageEditor/sidebar/AddToRecipeModal.tsx:66]: Let's define a constant for the new recipe id `@new`

File: src/pageEditor/sidebar/CreateRecipeModal.tsx [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+const CreateRecipeModal: React.VFC = () => {
+  const newRecipeIds = useSelector(selectNewRecipeIds);
+  const activeElement = useSelector(selectActiveElement);
+  const scope = useSelector(selectScope);
+  const [createRecipe] = useCreateRecipeMutation();
+  const create = useCreate();
+  const keepLocalCopy = useSelector(selectKeepLocalCopyOnCreateRecipe);
+
+  // TODO: This should be yup.SchemaOf<RecipeConfiguration> but we can't set the `id` property to `RegistryId`
```
Comment (BALEHOK) [inline src/pageEditor/sidebar/CreateRecipeModal.tsx:60]: This is just about types, isn't it?
Something like this?
```
yup.SchemaOf<RecipeConfiguration & { id: string }>
```

File: src/pageEditor/tabs/RecipeOptions.tsx [hunk 4]
```diff
@@ -52,19 +53,26 @@ const formRuntimeContext: RuntimeContext = {
   allowExpressions: false,
 };
 
-const RecipeOptions: React.VoidFunctionComponent = () => {
+const RecipeOptions: React.VFC = () => {
```
Comment (BALEHOK) [inline src/pageEditor/tabs/RecipeOptions.tsx:56]: Let's settle on the short version of the type then. I mean, from now on we always use the short version only.
Sounds good?

**Review by BALEHOK — APPROVED** [review]
Just a few minor comments. Good to merge.

File: src/pageEditor/panes/save/saveHelpers.ts (hunk 1 above)
Comment (BLoe) [reply src/pageEditor/panes/save/saveHelpers.ts:299]: https://en.wikipedia.org/wiki/Tabula_rasa

I can make the language more clear, it's kind of confusing, you're right.

File: src/pageEditor/sidebar/CreateRecipeModal.tsx (hunk 3 above)
Comment (BLoe) [reply src/pageEditor/sidebar/CreateRecipeModal.tsx:60]: Discussed offline, I'm going to remove this comment. I copied it from previous code and it's confusing and we don't understand what it's talking about anyway.

File: src/pageEditor/tabs/RecipeOptions.tsx (hunk 4 above)
Comment (BLoe) [reply src/pageEditor/tabs/RecipeOptions.tsx:56]: Yep, agree! We're going to have to update all of these for React 18 soon anyway with the changes to the `children` prop.

**Pull request merged**